### DISCUSSION TITLE
USB Serial protocol to play remotely with JS Web Serial API.

### DISCUSSION THREAD

**Discussion by nilzao** [darthcloud/BlueRetro#294] [Ideas]
Hi,

Thanks for sharing this project, I'm using it with my N64 and a generic BT gamepad.

TLDR: 
I need help to define/discuss a Serial protocol between browser and the ESP32 port (web browser gamepad api, and web serial api).

Long version:

I have a project/prototype running snes inside browser with webasm, based on xnes project.

With some mods, we are able to play 4 players remotely using html5, javascript with webrtc (streaming audio/video/gamepad commands), websocket (to make webrtc handshakes), js gamepad api (gamepads working inside browser).

Here's how it works today:

![P-FULL](https://user-images.githubusercontent.com/288571/172048708-f7916226-77a5-4b6b-8abb-a446e0d0d970.png)

1 browser (HOST-WRTC) running snes emulator with webrtc streaming media server, and webrtc gamepad bridge server (from webrtc to snes webasm input)
4 browsers are werbrtc streaming clients sending gamepad commands.

The new feature idea:
![P-HARDWARE](https://user-images.githubusercontent.com/288571/172049434-ff493276-1b75-4b86-bede-245937d8a9f5.png)

First problem:
- Define a standard protocol between Web Serial api using Uint8Array to send gamepad commands to esp32.
What send? How to send?

Something like:
[gamepad number] [btn1 stats] [btn2 stats] [axys1 position] ...
[0x01] [0x00] [0x00] [0x15A3] ...
[gamepad 1] [off] [off] [Float position]

Send bits instead of bytes, example 1 byte can send 8 buttons states.

With that, will be possible play ANY console remotely. I'm looking forward to play my N64 with my brother and friends across the country.

**Comment by darthcloud** [comment]
Sending button as byte is a big no no for me, big waste of space and additional latency for nothing.

The protocol format should be a packed version of BlueRetro generic_ctrl structure:
https://github.com/darthcloud/BlueRetro/blob/819efad1d98764bb45e58b300a26090564d641bb/main/adapter/adapter.h#L369

The buttons bitfield is 128 bits and value are define by this enum:
https://github.com/darthcloud/BlueRetro/blob/819efad1d98764bb45e58b300a26090564d641bb/main/adapter/adapter.h#L136

**Reply by darthcloud** [reply]
Forgot to leave link the actual gamepad buttons def
https://github.com/darthcloud/BlueRetro/blob/819efad1d98764bb45e58b300a26090564d641bb/main/adapter/adapter.h#L99

keyboard and gamepad bits overlap

**Reply by nilzao** [reply]
Nice one.

I will try gamepads first, keyboard later.

Here is a "hello world" with Xbox 360 buttons (should work with Xbox one/PS4/PS5 too)

https://github.com/hordeparty/prototype/tree/main/blueretro

Using this enum:
https://github.com/darthcloud/BlueRetro/blob/819efad1d98764bb45e58b300a26090564d641bb/main/adapter/adapter.h#L99

Mapped following this:
[BlueRetro v1.4+ mapping label reference](https://docs.google.com/spreadsheets/d/e/2PACX-1vT9rPK2__komCjELFpf0UYz0cMWwvhAXgAU7C9nnwtgEaivjsh0q0xeCEiZAMA-paMrneePV7IqdX48/pubhtml)

Can you evaluate this button "protocol" output please? I will build the axis protocol later.

**Reply by darthcloud** [reply]
I fully support what you want to do but I wont have time to make review or contribute sorry,

Ideally you would implement the BlueRetro thread that would handle the UART RX and and call adapter_bridge https://github.com/endpoint32/node0/blob/819efad1d98764bb45e58b300a26090564d641bb/main/adapter/adapter.c#L284

**Reply by nilzao** [reply]
No problem, thanks for helping.

The adapter.c adapter_bridge link:

https://github.com/darthcloud/BlueRetro/blob/819efad1d98764bb45e58b300a26090564d641bb/main/adapter/adapter.c#L284

I will work on that, cya.

**Comment by nilzao** [comment]
Hi @darthcloud,

I did a proof of concept, and it's working, I guess we can discuss the serial protocol now.

host.c and adapter.c mods:

```c
//host.c

/**
 * must add:
 * xTaskCreatePinnedToCore(&bt_serial_task, "bt_serial_task", 2048, NULL, 12, NULL, 0);
 * to int32_t bt_host_init(void) function;
 */
static void bt_serial_task(void *param) {
    while(1) {
		char command[8];
		if (fgets(command, sizeof(command), stdin)) {
			serial_bridge(command);
		}
        // I dunno what's going on, but it works for now:
        vTaskDelay(16 / portTICK_PERIOD_MS);
    }
}
```

```c
//adapter.c
/**
 * steps to reproduce:
 * build (sdk hw1), flash, monitor
 * pair some BT gamepad, wait to sync
 * send bytes to serial port
 */
void serial_bridge(char bytes[8]) {
	/*
	 *  serial_to_input changes:
	 *  ctrl_input->btns[0].value,
	 *  ctrl_input->axes[0].value and ctrl_input->axes[1].value
	 *  with some logic faking the BT gamepad that I have
	 */
	serial_to_input(bytes, ctrl_input);
	adapter_debug_print(ctrl_input); // printf debug
	if (wired_meta_init(ctrl_output)) {
		/* Unsupported system */
		return;
	}
	adapter_mapping(&config.in_cfg[0]);
	adapter_debug_print(&ctrl_output[0]); // printf debug
	wired_from_generic(config.out_cfg[0].dev_mode, &ctrl_output[0], &wired_adapter.data[0]);
}
```

My focus is gamepad 16 buttons, 5 axes, here's my first first version:

| | CMD/START | Controller idx | Btns stats | Axis0 XY | Axis1 XY | Axis2 XY | Axis3 XY | Axis4 XY | CRC/END |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| bits | 2 bits | 2 bits | 16 bits | 8 bits | 8 bits | 8 bits | 8 bits | 8 bits | 4 bits |


Total 64 bits, 8 bytes.

Some bandwidth calculation, running 12 controls at full capacity:

12 * 64 = 708 bits

2ms response: 1000 / 2 = 500

708 * 500 = 354.000bps

at 460800bps transfer rate, can run 12 controllers with 2ms latency or 6 controllers with 1ms latency (theorically).

Please share your considerations,

Thanks

**Comment by nilzao** [comment]
Hi again, I did a html5 page to debug and changed the protocol a little:

| | CMD/START | Controller idx | Btns stats | Axis0 XY | Axis1 XY | Axis2 XY | Axis3 XY | CRC/END |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | 
| bits | 4 bits | 4 bits | 16 bits | 8 bits | 8 bits | 8 bits | 8 bits | 8 bits |

Using [W3C Gamepad](https://w3c.github.io/gamepad/#fig-visual-representation-of-a-standard-gamepad-layout) now (except button 16).

https://htmlpreview.github.io/?https://github.com/hordeparty/prototype/blob/main/blueretro/index.html

I forked the BlueRetro project and started some mods, I'm getting there.
